tests/src/test/scala/org/http4s/UriSpec.scala

```diff
@@ -1349,6 +1349,59 @@ class UriSpec extends Http4sSuite {
       assertEquals(pre.concat(post), path1)
     }
 
+    test("empty concat 1")(assertEquals(path"".concat(path""), path""))
+    test("empty concat 2")(assertEquals(path"/".concat(path""), path"/"))
+    test("empty concat 3")(assertEquals(path"".concat(path"/"), path"/"))
+    test("empty concat 4")(assertEquals(path"/".concat(path"/"), path"/"))
+
+    test("simple concat 1")(assertEquals(path"a/".concat(path""), path"a/"))
+    test("simple concat 2")(assertEquals(path"a".concat(path"/"), path"a/"))
+    test("simple concat 3")(assertEquals(path"".concat(path"/a"), path"/a"))
+    test("simple concat 4")(assertEquals(path"/".concat(path"a"), path"/a"))
+
+    test("When left side of concat is absolute then the result is absolute") {
+      forAll((left: Path, right: Path) => assert(left.toAbsolute.concat(right).absolute))
+    }
+
+    test("When right side of concat ends with slash then the result ends with slash") {
+      forAll((left: Path, right: Path) => assert(left.concat(right.addEndsWithSlash).endsWithSlash))
+    }
+
+    test("splitAt(0)._2 is identity") {
+      forAll((path: Path) => assertEquals(path.splitAt(0)._2, path))
+    }
+
+    test("splitAt(-1) should equal splitAt(0)") {
+      forAll((path: Path) => assertEquals(path.splitAt(-1), path.splitAt(0)))
+    }
+
+    test("splitAt(segments.size + 1)._1 is identity") {
```

Comment:
What about `splitAt(segments.size)._1`?

Reply:
`splitAt(segments.size)._1` is not an identity. But let me explain this topic.

When we split a path in the middle (between some segments) then we split it at a slash and we need to decide what will happen with that slash after splitting. Let's say we split `path"/a/b/"` at 1. The result could be:
-  `(path"/a", path"b/")` - none of the sides receives the slash
-  `(path"/a/", path"b/")` - left side receives the slash
-  `(path"/a", path"/b/")` - right side receives the slash
-  `(path"/a/", path"/b/")` - both sides receive the slash
As I said before "I did not know how exactly Path should behave" and this applies here as well. From my perspective the first option is wrong (some laws would break in some cases). The remaining 3 options are fine. The "both" sound best as it is symmetric. But other than that, "left" and "right" will work. And what we used to have was the "right" option, and I stayed with that, as it minimised potential for breaking something.

So what happens if we extrapolate this behaviour to `splitAt(segments.size)`? We end up with `path"/a/b/".splitAt(2) == (path"/a/b", path"/")`, so `splitAt(segments.size)._1` is not an identity.

I added example-based tests which will document and freeze this specific behaviour. Mentally I put these tests in a different category than the properties, in the sense that the properties document something that fundamentally makes sense, whereas these examples (and the lack of a "splitAt(segments.size)._1 is identity") document a design decision.

Reply:
Thank you, this is great. Sorry, I have one more question.

> `(path"/a", path"/b/")` - right side receives the slash

In this case, the right path becomes an absolute path, correct? I don't disagree with preserving this behavior for compatibility, but personally this strikes me as very odd.

Reply:
Indeed. It's confusing because, after all the `/b/` is relative to `/a`. 

One of the reasons why it was design like this, is that http4s's DSL expects all paths to be absolute (`case Root ... => ...`). That makes sense at first sight, because Requests are required to use absolute paths, but then you realise things like Routers, servletContext etc exist.

You can see that in order to have a `splitAt` that behaves in a reasonable way, and at the same time to satisfy the DSL, I had to do a `(l.toAbsolute, r.toAbsolute)` in `Message`.

I'm open to discussing changing this behaviour.

Reply:
Ah, that's right, the DSL. Thanks for the excellent explanations. I think what you've done here is very good. At the least, because it specifies, documents, and freezes the behavior.

I'll open an issue about the absolute path thing, we can consider for 1.0 maybe.

Reply:
https://github.com/http4s/http4s/issues/5798

Reply:
Just to be clear about it. I think it would be possible to have the 

> (path"/a/", path"b/") - left side receives the slash

or the

> (path"/a/", path"/b/") - both sides receive the slash

behaviour, without changing how the DSL works (because we make the DSL happy with the `(l.toAbsolute, r.toAbsolute)` anyway). It's actually the `Message#scriptName` (the left side of the split) receiveing a slash at the end that will cause more hard-to-foresee problems.

Reply:
What I meant to say is that the original (current seriess/0.22) design give the slash to the right-hand-side because that's what the DSL needs. In this PR this is no longer needed because the `r.toAbsolute` makes the DSL happy. But I kept this design-decision to avoid breaking something elsewhere (other than the DSL).